crates/matrix-sdk-base/src/room/tombstone.rs

```diff
@@ -12,18 +12,291 @@
+    /// —i.e. the room replacing this one.
+    ///
+    /// A room is tombstoned if it has received a [`m.room.tombstone`] state
+    /// event; see [`Room::tombstone_content`].
+    ///
+    /// [`m.room.tombstone`]: https://spec.matrix.org/v1.14/client-server-api/#mroomtombstone
+    pub fn successor_room(&self) -> Option<SuccessorRoom> {
+        self.tombstone_content().map(|tombstone_event| SuccessorRoom {
+            room_id: tombstone_event.replacement_room,
+            reason: tombstone_event.body.is_empty().not().then_some(tombstone_event.body),
+        })
+    }
+
+    /// If this room is the successor of a tombstoned room, return the
+    /// “reference” to the predecessor room.
+    ///
+    /// A room is tombstoned if it has received a [`m.room.tombstone`] state
+    /// event; see [`Room::tombstone_content`].
+    ///
+    /// To determine if a room is the successor of a tombstoned room, the
+    /// [`m.room.create`] must have been received, **with** a `predecessor`
+    /// field. See [`Room::create_content`].
+    ///
+    /// [`m.room.tombstone`]: https://spec.matrix.org/v1.14/client-server-api/#mroomtombstone
+    /// [`m.room.create`]: https://spec.matrix.org/v1.14/client-server-api/#mroomcreate
+    pub fn predecessor_room(&self) -> Option<PredecessorRoom> {
+        self.create_content().and_then(|content_event| content_event.predecessor).map(
+            |predecessor| PredecessorRoom {
+                room_id: predecessor.room_id,
+                event_id: predecessor.event_id,
+            },
+        )
+    }
+}
+
+/// When a room A is tombstoned, it is replaced by a room B. The room A is the
+/// predecessor of B, and B is the successor of A. This type holds information
+/// about the successor room. See [`Room::successor_room`].
+///
+/// A room is tombstoned if it has received a [`m.room.tombstone`] state event.
+///
+/// [`m.room.tombstone`]: https://spec.matrix.org/v1.14/client-server-api/#mroomtombstone
+#[derive(Debug)]
+pub struct SuccessorRoom {
+    /// The ID of the next room replacing this (tombstoned) room.
+    pub room_id: OwnedRoomId,
+
+    /// The reason why the room has been tombstoned.
+    pub reason: Option<String>,
+}
+
+/// When a room A is tombstoned, it is replaced by a room B. The room A is the
+/// predecessor of B, and B is the successor of A. This type holds information
+/// about the predecessor room. See [`Room::predecessor_room`].
+///
+/// To know the predecessor of a room, the [`m.room.create`] state event must
+/// have been received.
+///
+/// [`m.room.tombstone`]: https://spec.matrix.org/v1.14/client-server-api/#mroomtombstone
+/// [`m.room.create`]: https://spec.matrix.org/v1.14/client-server-api/#mroomcreate
+#[derive(Debug)]
+pub struct PredecessorRoom {
+    /// The ID of the old room.
+    pub room_id: OwnedRoomId,
+
+    /// The event ID of the last known event in the predecesssor room.
+    pub event_id: OwnedEventId,
```

Comment:
Could we name this `last_event_id`? Otherwise it's very unclear what event id this is…

Reply:
Sure! Good idea.